rust/datafusion/src/sql/planner.rs

```diff
@@ -894,6 +908,131 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+            )));
+        }
+
+        if last_field.is_some() {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Unsupported Interval Expression with last_field {:?}",
+                last_field
+            )));
+        }
+
+        if fractional_seconds_precision.is_some() {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Unsupported Interval Expression with fractional_seconds_precision {:?}",
+                fractional_seconds_precision
+            )));
+        }
+
+        // Should we resort parts on overflow?
+        let resort_parts = |days: i32, seconds: f32| -> (i32, f32) { (days, seconds) };
+
+        let calculate_from_part =
+            |interval_period_str: &str, interval_type: &str| -> Result<(i32, f32)> {
+                let interval_period = match interval_period_str.parse::<f32>() {
+                    Ok(n) => n,
+                    Err(_) => {
+                        return Err(DataFusionError::SQL(ParserError(format!(
+                            "Unsupported Interval Expression with value {:?}",
+                            value
+                        ))))
+                    }
+                };
+
+                if interval_period > (i32::MAX as f32) {
+                    return Err(DataFusionError::NotImplemented(format!(
+                        "Interval field value out of range: {:?}",
+                        value
+                    )));
+                }
+
+                const SECONDS_PER_HOUR: f32 = 3_600_f32;
+                const MILLIS_PER_SECOND: f32 = 1_000_f32;
+
+                match interval_type.to_lowercase().as_str() {
+                    "year" => Ok(((interval_period * 365_f32) as i32, 0.0)),
+                    "month" => Ok(((interval_period * 30_f32) as i32, 0.0)),
+                    "day" | "days" => Ok((interval_period as i32, 0.0)),
+                    "hour" | "hours" => Ok(resort_parts(
+                        0,
+                        interval_period * SECONDS_PER_HOUR * MILLIS_PER_SECOND,
+                    )),
+                    "minutes" | "minute" => Ok(resort_parts(
+                        0,
+                        interval_period * 60_f32 * MILLIS_PER_SECOND,
+                    )),
+                    "seconds" | "second" => {
+                        Ok(resort_parts(0, interval_period * MILLIS_PER_SECOND))
+                    }
+                    "milliseconds" | "millisecond" => Ok((0, interval_period)),
+                    _ => {
+                        return Err(DataFusionError::NotImplemented(format!(
+                            "Invalid input syntax for type interval: {:?}",
+                            value
+                        )))
+                    }
+                }
+            };
+
+        let mut parts = value.split_whitespace();
+        let mut result_days: i64 = 0;
+        let mut result_millis: i64 = 0;
+
+        loop {
```

Comment:
If you wanted to make this more "idomatic" rust you could write something like:

```
for interval_period_str in parts {
...
}
```

And if you really wanted to get your functional programming groove on you could do something like this with `fold` (untested):

```
let (result_days, result_millis) = value.split_whitespace()
  .fold((0,0), |(result_days, result_millis),interval_period_string| {
    // calculate the diff
    (result_days + diff_days, result_millis + diff_millis)
});
```

I don't think there is any real performance difference but I figured I would point it out if you wanted to try

Reply:
Awesome, thank you! I will try.

Reply:
So, there is a restriction to use it:

```
1084 |         let mut parts = value.split_whitespace();
     |             --------- move occurs because `parts` has type `SplitWhitespace<'_>`, which does not implement the `Copy` trait
1085 | 
1086 |         for interval_period_str in parts {
     |                                    -----
     |                                    |
     |                                    `parts` moved due to this implicit call to `.into_iter()`
     |                                    help: consider borrowing to avoid moving into the for loop: `&parts`
...
1089 |                 parts.next().unwrap_or("second"),
     |                 ^^^^^ value borrowed here after move
```

![image](https://user-images.githubusercontent.com/572096/106817177-90bf3400-6687-11eb-897c-e180ca156afc.png)

It's needed to support

```
SELECT (INTERVAL '1')
SELECT (ITNERVAL '1 day 1')
```

Reply:
@ovr  -- I see what you are saying (that the iterator may need to consume an extra item during the loop). Thank you for trying -- I am sorry if you wasted much time on that